crates/extensions/c8y_mapper_ext/src/compatibility_adapter.rs

```diff
@@ -0,0 +1,166 @@
+        ["tedge", "commands", "res", "control", "restart"] => {
+            convert_from_old_agent_response("restart", payload)
+        }
+        ["tedge", "commands", "res", "software", "list"] => {
+            convert_from_old_agent_response("software_list", payload)
+        }
+        ["tedge", "commands", "res", "software", "update"] => {
+            convert_from_old_agent_response("software_update", payload)
+        }
+        [_, "device", "main", "", "", "cmd", "restart", cmd_id] => {
+            convert_to_old_agent_request("control/restart", cmd_id, payload)
+        }
+        [_, "device", "main", "", "", "cmd", "software_list", cmd_id] => {
+            convert_to_old_agent_request("software/list", cmd_id, payload)
+        }
+        [_, "device", "main", "", "", "cmd", "software_update", cmd_id] => {
+            convert_to_old_agent_request("software/update", cmd_id, payload)
+        }
+        _ => Ok(None),
+    }
+}
+
+fn convert_to_old_agent_request(
+    cmd_type: &str,
+    cmd_id: &str,
+    payload: &[u8],
+) -> Result<Option<MqttMessage>, String> {
+    if payload.is_empty() {
+        // Ignore clearing message
+        return Ok(None);
+    }
+    if let Ok(Value::Object(mut request)) = serde_json::from_slice(payload) {
+        if let Some(Value::String(status)) = request.get("status") {
+            if status != "init" {
+                // Ignore non-init state
+                return Ok(None);
+            }
+            request.insert("id".to_string(), Value::String(cmd_id.to_string()));
+            request.remove("status"); // as the old agent denies unknown fields
+            if let Ok(updated_payload) = serde_json::to_vec(&Value::Object(request)) {
+                if let Ok(topic) = Topic::new(&format!("tedge/commands/req/{cmd_type}")) {
+                    return Ok(Some(
+                        MqttMessage::new(&topic, updated_payload).with_qos(QoS::AtLeastOnce),
+                    ));
+                }
+            }
+        }
+    }
+
+    Err(format!(
+        "Fail to inject command 'id' into agent {cmd_type} request: {}",
+        std::str::from_utf8(payload).unwrap_or("non utf8 payload")
+    ))
+}
+
+fn convert_from_old_agent_response(
+    cmd_type: &str,
+    payload: &[u8],
+) -> Result<Option<MqttMessage>, String> {
+    if let Ok(Value::Object(response)) = serde_json::from_slice(payload) {
+        if let Some(Value::String(cmd_id)) = response.get("id") {
+            if let Ok(topic) = Topic::new(&format!("te/device/main///cmd/{cmd_type}/{cmd_id}")) {
+                return Ok(Some(
+                    MqttMessage::new(&topic, payload)
+                        .with_retain()
+                        .with_qos(QoS::AtLeastOnce),
+                ));
+            }
+        }
+    }
+
+    Err(format!(
+        "Fail to extract command 'id' from agent {cmd_type} response: {}",
+        std::str::from_utf8(payload).unwrap_or("non utf8 payload")
+    ))
+}
```

Comment:
I'd add some unit test coverage, **if there's time**, especially for the `convert_to_old_agent_request` function which has some deep nested logic.